bson/bsoncodec/default_value_decoders.go

```diff
@@ -455,12 +486,12 @@ func (DefaultValueDecoders) UndefinedDecodeValue(dc DecodeContext, vr bsonrw.Val
 		return ValueDecoderError{Name: "UndefinedDecodeValue", Types: []reflect.Type{tUndefined}, Received: val}
 	}
 
-	if vr.Type() != bsontype.Undefined {
+	if vr.Type() != bsontype.Undefined && vr.Type() != bsontype.Null {
 		return fmt.Errorf("cannot decode %v into an Undefined", vr.Type())
 	}
 
 	val.Set(reflect.ValueOf(primitive.Undefined{}))
-	return vr.ReadUndefined()
+	return vr.Skip()
```

Comment:
I think this is valid and made the same change in `MinKeyDecodeValue` and `MaxKeyDecodeValue`. `ReadUndefined` doesn't return anything, so we should be able to just skip the value.

Reply:
Turns out this is not valid. It's fine for BSON, but extended JSON requires undefined, max key, and min key values to be in a specific form (e.g. min key must be `{$minKey: 1}` and something like `{$minKey: 2}` is supposed to fail) so we have to call the correct `Read*` methods on `ValueReader` to make sure the validation is done.